nilearn/decomposition/base.py

```diff
@@ -0,0 +1,453 @@
+    else:
+        if reduction_ratio is None:
+            reduction_ratio = 1
+        else:
+            reduction_ratio = float(reduction_ratio)
+        if not 0 <= reduction_ratio <= 1:
+            raise ValueError('Reduction ratio should be between 0. and 1.,'
+                             'got %.2f' % reduction_ratio)
+
+    if confounds is None:
+        confounds = itertools.repeat(confounds)
+
+    # Precomputing number of samples for preallocation
+    subject_n_samples = np.zeros(len(imgs), dtype='int64')
+    for i, img in enumerate(imgs):
+        this_n_samples = img.shape[3]
+        if reduction_ratio == 'auto':
+            subject_n_samples[i] = min(n_components,
+                                       this_n_samples)
+        else:
+            subject_n_samples[i] = int(ceil(this_n_samples *
+                                            reduction_ratio))
+
+    data_list = Parallel(n_jobs=n_jobs)(
+        delayed(_mask_and_reduce_single)(
+            masker,
+            img, confound,
+            n_samples,
+            memory=memory,
+            memory_level=memory_level,
+            random_state=random_state
+        ) for img, confound, n_samples in zip(imgs, confounds,
+                                              subject_n_samples))
+    n_samples = np.sum(subject_n_samples)
+    n_voxels = np.sum(_safe_get_data(masker.mask_img_))
+    data = np.empty((n_samples, n_voxels), order='F',
+                    dtype='float64')
+
+    current_position = 0
+    for i, next_position in enumerate(np.cumsum(subject_n_samples)):
+        data[current_position:next_position] = data_list[i]
+        current_position = next_position
+    return data
+
+
+def _mask_and_reduce_single(masker,
+                            img, confound,
+                            n_samples,
+                            memory=None,
+                            memory_level=0,
+                            random_state=None):
+    """Utility function for multiprocessing from MaskReducer"""
+    this_data = masker.transform(img, confound)
+    random_state = check_random_state(random_state)
+
+    if n_samples <= this_data.shape[0] // 4:
+        U, S, _ = cache(randomized_svd, memory,
+                        memory_level=memory_level,
+                        func_memory_level=3)(this_data.T,
+                                             n_samples,
+                                             transpose=True,
+                                             random_state=random_state)
+        U = U.T
+    else:
+        U, S, _ = cache(linalg.svd, memory,
+                        memory_level=memory_level,
+                        func_memory_level=3)(this_data.T,
+                                             full_matrices=False)
+        U = U.T[:n_samples].copy()
+        S = S[:n_samples]
+    U = U * S[:, np.newaxis]
+    return U
+
+
+class BaseDecomposition(BaseEstimator, CacheMixin):
+    """Base class for decomposition estimator.
```

Comment:
Suggestion for higher specificity: "for MATRIX decomposition"

Also "estimatorS"